oxenc/common.h

```diff
@@ -34,22 +35,47 @@ concept tuple_like = requires { std::tuple_size<Tuple>::value; };
 
 // True if the type is a std::string, std::string_view, or some a basic_string<Char> for some
 // single-byte type Char.
-template <typename T>
-constexpr bool is_string_like = false;
-template <typename Char>
-inline constexpr bool is_string_like<std::basic_string<Char>> = sizeof(Char) == 1;
-template <typename Char>
-inline constexpr bool is_string_like<std::basic_string_view<Char>> = sizeof(Char) == 1;
+template <typename T, typename U = T::value_type>
+concept string_like = basic_char<U> &&
```

Comment:
Okay maybe this is a dumb question, but isn't the purpose of the template specialization (now a concept) to be:
```
// True if the type is a std::string, std::string_view, or some a basic_string<Char> for some
// single-byte type Char.
```
So wouldn't it make sense that if we want it constrain to string/view of some Char type, we could assume that there is a `::value_type`?

Reply:
Also, don't spots like [here](https://github.com/oxen-io/oxen-encoding/blob/dev/oxenc/bt_serialize.h#L298) assume a valid `::value_type`?

Reply:
> So wouldn't it make sense that if we want it constrain to string/view of some Char type, we could assume that there is a ::value_type?

We currently use `is_string_like` in a couple `if constexpr`s in the code, which means dropping `is_string_like` in favour of just the concept without any backing template means the concept needs to be able to return true/false for any generic type, but currently it's true/compilation error.  You had to add `value_type` typedefs in a couple places to work around that compilation error, but those are workarounds for a limitation of the concept that didn't exist previously.

Having a `string_like` concept is perfectly fine, but trying to do it without partial specialization doesn't seem workable.  This is a simple and robust pattern:

```C++
template <typename T>
inline constexpr bool is_whatever = false;
template <typename T>
inline constexpr bool is_whatever<Whatever<T>> = true;

template <typename T>
concept whatever = is_whatever<T>;
```

because concepts do not allow, by design, template specialization: once your type requirements get complicated enough to warrant it, concept + specialization does the job perfectly well, keeping `concept` quite a bit simpler by not having any specialization logic in it.

There's a separate but related problem here, though, and it is that the negative path of the concept as written will instantiating the `std::string_view<T>/std::string_view<T>` just to evaluate the concept (which is likely a libc++ 19 issue).  But it's also forcing the compiler to "see" an instantiation that we may not care about or ever use, because if I have something like a `std::vector<int>` that passes through a `if constexpr (string_like<T>)` then it will compile (because there is a `::value_type`) but won't enter, but the still had the instantiate the `std::basic_string<int>` and `std::basic_string_view<int>`  classes along the way just to be able to evaluate it.

With template specialization that never happens: because `if constexpr (string_like<T>)` is simply false because `vector<int>` doesn't find any specialization and so goes straight to the `false` fallback.

There are other ways you might go about solving it, but I think it's pretty much always going to involve template specialization because what we're trying to match here is "any basic_string<T> or basic_string_view<T> that has a size-1 T", and that's basically a template matching problem.

Reply:
> Also, don't spots like [here](https://github.com/oxen-io/oxen-encoding/blob/dev/oxenc/bt_serialize.h#L298) assume a valid `::value_type`?

Yeah, that looks wrong too; that one is definitely on me from embracing concepts perhaps a bit over-zealously.

There *is* a value way to check on a `::value_type` with a `requires (T) { ... }`, because a `requires` converts does-not-compile into false, and so some cases might be fixable with that, but the string_like I'm thinking is still going to need some specialization.

Reply:
okay, makes sense. thank you for the writeup. changed after chatting off-band as well